Add derive macros for `FromValue` and `IntoValue` to ease the use of `Value`s in Rust code

cptpiepmatz opened the pull request:
<!--
if this PR closes one or more issues, you can automatically link the PR with
them by using one of the [*linking keywords*](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword), e.g.
- this PR should close #xxxx
- fixes #xxxx

you can also mention related issues, PRs or discussions!
-->

# Description
<!--
Thank you for improving Nushell. Please, check our [contributing guide](../CONTRIBUTING.md) and talk to the core team before making major changes.

Description of your pull request goes here. **Provide examples and/or screenshots** if your changes affect the user experience.
-->
After discussing with @sholderbach the cumbersome usage of `nu_protocol::Value` in Rust, I created a derive macro to simplify it. I’ve added a new crate called `nu-derive-value`, which includes two macros, `IntoValue` and `FromValue`. These are re-exported in `nu-protocol` and should be encouraged to be used via that re-export.

The macros ensure that all types can easily convert from and into `Value`. For example, as a plugin author, you can define your plugin configuration using a Rust struct and easily convert it using `FromValue`. This makes plugin configuration less of a hassle.

I introduced the `IntoValue` trait for a standardized approach to converting values into `Value` (and a fallible variant `TryIntoValue`). This trait could potentially replace existing `into_value` methods. Along with this, I've implemented `FromValue` for several standard types and refined other implementations to use blanket implementations where applicable.

I made these design choices with input from @devyn.

There are more improvements possible, but this is a solid start and the PR is already quite substantial.
# User-Facing Changes
<!-- List of all changes that impact the user experience here. This helps us keep track of breaking changes. -->
For `nu-protocol` users, these changes simplify the handling of `Value`s. There are no changes for end-users of nushell itself.

# Tests + Formatting
<!--
Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used` to check that you're using the standard code style
- `cargo test --workspace` to check that all tests pass (on Windows make sure to [enable developer mode](https://learn.microsoft.com/en-us/windows/apps/get-started/developer-mode-features-and-debugging))
- `cargo run -- -c "use toolkit.nu; toolkit test stdlib"` to run the tests for the standard library

> **Note**
> from `nushell` you can also use the `toolkit` as follows
> ```bash
> use toolkit.nu  # or use an `env_change` hook to activate it automatically
> toolkit check pr
> ```
-->
Documenting the macros itself is not really possible, as they cannot really reference any other types since they are the root of the dependency graph. The standard library has the same problem ([std::Debug](https://doc.rust-lang.org/stable/std/fmt/derive.Debug.html)). However I documented the `FromValue` and `IntoValue` traits completely.

For testing, I made of use `proc-macro2` in the derive macro code. This would allow testing the generated source code. Instead I just tested that the derived functionality is correct. This is done in `nu_protocol::value::test_derive`, as a consumer of `nu-derive-value` needs to do the testing of the macro usage. I think that these tests should provide a stable baseline so that users can be sure that the impl works.

# After Submitting
<!-- If your PR had any user-facing changes, update [the documentation](https://github.com/nushell/nushell.github.io) after the PR is merged, if necessary. This will help us keep the docs up to date. -->
With these macros available, we can probably use them in some examples for plugins to showcase the use of them.

sholderbach commented on crates/nu-protocol/src/value/test_derive.rs:
```diff
@@ -0,0 +1,375 @@
+        ]
+    }
+
+    fn value() -> Value {
+        Value::test_list(vec![
+            Value::test_record(record! {
+                "type" => Value::test_string("unit")
+            }),
+            Value::test_record(record! {
+                "type" => Value::test_string("tuple"),
+                "content" => Value::test_list(vec![
+                    Value::test_int(12),
+                    Value::test_string("Tuple variant")
+                ])
+            }),
+            Value::test_record(record! {
+                "type" => Value::test_string("struct"),
+                "content" => Value::test_record(record! {
+                    "a" => Value::test_int(34),
+                    "b" => Value::test_string("Struct variant")
+                })
+            }),
+        ])
+    }
```
Q: should we provide a derive for `enum` if we don't have a canonical representation for that in the Nushell type system (yet) and we have to make one up here for completeness sake.

We kind of have enum-like patterns with options that only accept a set of strings, they could map to C-style enums. This at the same time would have a distinct encoding from a tagged-union as suggested here.

cptpiepmatz replied:
We can, at least for now, drop the derive for enums and let users handle enums themselves, as they can also implement `FromValue` and `IntoValue` on their own. But it would be still very practical. Also `Option`s are kind of annoying since no `Value` variant depicts that a value is optional.

sholderbach replied:
As we don't make a non-null guarantee for outputs at the moment I think your `Option` implementation is good.
But yeah we make that distinction for command parameters (and field access via the optional/standard cell-path).

sholderbach reviewed (commented):
Thanks for diving into the proc-macro fun world for this.

As this touches key aspects with regards to representation-stability, I am inclined to make sure we only provide mappings if we have unambiguous, sane semantics for the particular types.

cptpiepmatz commented on crates/nu-protocol/src/value/from_value.rs:
```diff
@@ -1,353 +1,499 @@
+    /// The operation might fail if the `Value` contains unexpected types or
+    /// structures.
+    ///
+    /// The `call_span` parameter is used to trace errors when they are not
+    /// directly related to the `Value` itself.
+    /// For example, when using [`Value::Record`] and a field is missing,
+    /// nushell typically tracks errors back to user input or files.
+    /// However, the specific column name triggering the error might not be part
+    /// of the direct user input, as the column name could be embedded within
+    /// the binary itself.
+    /// By using `call_span`, you can link the error to the broader command or
+    /// input that triggered the function call, helping identify where the issue
+    /// with a missing field (or similar problem) began.
+    fn from_value(v: Value, call_span: Span) -> Result<Self, ShellError>;
```
I also kind of dislike that `call_span` as we don't need it most of the time. What other `ShellError` could we use or should I make a new variant to catch this kind of error? The `span` for `ShellError::CantFindColumn` would need to refer "into the binary". That's why @devyn suggested the `call_span` here. I'm up for avoiding this `call_span`.

cptpiepmatz commented on crates/nu-protocol/src/value/from_value.rs:
```diff
@@ -520,60 +626,39 @@ impl FromValue for Spanned<PathBuf> {
-                help: None,
-            }),
-        }
+impl<T> FromValue for Spanned<T>
+where
+    T: FromValue,
+{
+    fn from_value(v: Value, call_span: Span) -> Result<Self, ShellError> {
+        let span = v.span();
+        Ok(Spanned {
+            item: T::from_value(v, call_span)?,
+            span,
+        })
```
I'm not sure what I should make from this comment. Is this blanket implementation an issue or should we avoid Rust primitives?

sholderbach replied:
I think we should have both. 

Just spitballing that we could add an additional trait for just getting `Spanned<T>` and using that in the `nu_engine::prelude` and the `CallExt` trait.

cptpiepmatz commented:
@sholderbach In general, are the primitives implementations fine?

cptpiepmatz commented:
About the `call_span`. I opted to remove the `call_span` from the `FromValue::from_value` signature and removed all changes I made that added that second parameter. Since we have now no span anymore for the `CantFindColumn` variant, I made the `span` an `Option<Span>` to allow leaving that span out. To still have the column name in the error message, I added it in the main error message.

Also I removed `IntoValue` for `u64` and `usize`. 

I'm not sure what we do about enums now, having them in some way would be a major benefit, at least enums with only unit variants should be supported in some way.

devyn commented:
This is really awesome, great work @cptpiepmatz!

devyn commented:
> I'm not sure what we do about enums now, having them in some way would be a major benefit, at least enums with only unit variants should be supported in some way.

I think maybe we should just do that for now, and gradually add renaming and tuple/struct variants over time, make it basically like serde. But just literally the string of the enum variant would be ok for now.

sholderbach reviewed (commented):
Thanks for resolving the `call_span` part and the truncating ints. 
I concur with @devyn on the enum front, that we could possibly roll out the basic variants for now. If think we can build out the creature comforts later as needed (like special annotation to make variants transparent, proc macro error handling etc.). 
And direct tagged union support may depend a bit on Nushell language evolution.

I am excitedk if we can use the new patterns in a few place inside the core code base.

cptpiepmatz commented:
I realized that `nu-derive-value` is very specific. It wouldn't really fit for other proc macros if someone would ever want to implement something else. Should I name it `nu-derive` instead? 🤔

cptpiepmatz commented:
I pulled the meta implementation for enums out and replaced it with `Value::String` representations for plain enums and a compile time error if tried on more complex enums.

The naming for these values is done like before but can now be configured using `#[nu_value(rename_all = "...")]` on an enum.

I also added some more tests, they include:
 - derived enums rename properly
 - incorrectly used attributes are denied
 - complex enums and unions are denied

To work this out smoothly, I shoved some bits around in the derive crate.

Also I updated the docs for `IntoValue` and `FromValue` to explain that enum behavior. For that I pulled `convert_case` as a dependency into `nu_protocol` because `cargo doc` doesn't use dev-dependencies. But I figured that the link is useful and you probably already built `convert_case` as it is a dependency for the macros.

IanManske reviewed (commented):
Thanks for the PR! I only looked at it a little bit (I'll leave it to Devyn and Stefan to review in more depth), but just a few comments:

cptpiepmatz commented on crates/nu-protocol/src/value/into_value.rs:
```diff
@@ -0,0 +1,206 @@
+    /// Converts the given value to a [`Value`].
+    fn into_value(self, span: Span) -> Value;
+
+    /// Converts the given value to a [`Value`] with an unknown [`Span`].
+    ///
+    /// Internally this simply calls [`Span::unknown`] for the `span`.
+    fn into_value_unknown(self) -> Value {
+        Self::into_value(self, Span::unknown())
+    }
```
Yeah, I think you're right. The only real use-case for this function should be some constants the engine provides directly or tests. And those constants are the minority.

cptpiepmatz commented on crates/nu-protocol/src/value/into_value.rs:
```diff
@@ -0,0 +1,206 @@
+    }
+}
+
+macro_rules! tuple_into_value {
```
I see your point but that is the same concern for all tuples in Rust. I think users should be allowed to decide for themselves. When `nu_protocol` doesn't implement them, no one will be able to use tuples because of the orphan rule.

IanManske replied:
Fair enough, I guess we can keep the tuple implementations in case someone else needs them.

cptpiepmatz commented on crates/nu-protocol/src/value/into_value.rs:
```diff
@@ -0,0 +1,206 @@
+primitive_into_value!(u8 as i64, int);
+primitive_into_value!(u16 as i64, int);
+primitive_into_value!(u32 as i64, int);
+// u64 and usize may be truncated as Value only supports i64.
```
Done.

devyn commented on crates/nu-derive-value/src/into.rs:
Same here, this is not so complicated as `from` but it would still be nice to see a few more comments just laying out the process from a high-level perspective

devyn commented on crates/nu-derive-value/src/tests.rs:
Great job with the tests, looks like really good coverage here. Thanks!

devyn reviewed (changes requested):
Thanks @cptpiepmatz! I think this mostly in really good shape, just a couple of minor things I'd like to see and then unless @sholderbach has some more comments I think we can probably merge it.

cptpiepmatz commented on crates/nu-derive-value/src/attributes.rs:
```diff
@@ -0,0 +1,112 @@
+                match ident.to_string().as_str() {
+                    "rename_all" => {
+                        let case: LitStr = meta.value()?.parse()?;
+                        let case = match case.value().as_str() {
```
> I think we should try to follow the serde names. Maybe `lower` should be `lower with spaces`, `UPPER` be `UPPER WITH SPACES`, then the flat ones should be `lowercase` and `UPPERCASE`

I kept all the names I currently have as they are names of `convert_case`. But because all names are case sensitive and act as kind of a template, I added all `serde` names for the cases.

> I'm not sure what to call `Train-Case`, I've never heard it called this but agree it's useful since PowerShell uses it.

It's just the name that `convert_case` uses. In the generated documenation of the traits `IntoValue` and `FromValue` this is also referenced, so this should be fine.

> And I hate to be no-fun but we probably don't need tOGGLE and aLtErNaTiNg, I just can't imagine anyone wanting that in this context. 😛 Keep in mind that anything we add here, we're supporting as a public interface.

Yeah, not really useful. I just implemented all of them but now removed these two.

cptpiepmatz commented on crates/nu-derive-value/src/attributes.rs:
```diff
@@ -0,0 +1,112 @@
+                            "snake_case" => Case::Snake,
+                            "UPPER_SNAKE_CASE" | "SCREAMING_SNAKE_CASE" => Case::UpperSnake,
+                            "kebab-case" => Case::Kebab,
+                            "COBOL-CASE" | "UPPER-KEBAB-CASE" => Case::Cobol,
```
I added that as an alias.

cptpiepmatz commented on crates/nu-derive-value/src/from.rs:
I tried to document the functions pretty extensively now. All of them explain what code they produce on a high level and give examples how the generate code looks like. I also documented the crate itself.

But I disabled `doctest` for the entire crate as they don't really work in this crate. I tried to use the functions in the crate and check for equality but doctests use the crate as a consumer and then you only get the two macros.

devyn commented:
Thank you @cptpiepmatz! This is a great thing to have crossed off the list, and will help with making command and plugin code a lot easier to write.

fdncred commented:
Looks like there may be issues with other platforms. https://github.com/nushell/nushell/actions/runs/9556813143/job/26342820384
